File: domain/vocabulary_service.py
"""Pure vocabulary service - no UI dependencies.

Manages vocabulary CRUD operations with proper validation and error handling.
"""

from __future__ import annotations
from typing import Callable, Any
from dataclasses import dataclass

from domain.exceptions import (
    DuplicateEntryError,
    JyutpingValidationError,
    ValidationError,
)
from domain.jyutping_validation import validate_jyut_syllables
from domain.duplicate_rules import is_duplicate_jy, is_exact_duplicate_entry


@dataclass(frozen=True)
class VocabEntry:
    """Immutable vocabulary entry."""
    jyutping: str
    hanzi: str
    meanings: list[str]
    categories: list[str]
    notes: str = ""

    def to_dict(self) -> dict[str, Any]:
        """Export to dictionary format."""
        return {
            "jyutping": self.jyutping,
            "hanzi": self.hanzi,
            "meaning": ", ".join(self.meanings),
            "gloss": ", ".join(self.meanings),  # Legacy alias
            "categories": list(self.categories),
            "category": self.categories[0] if self.categories else "",
            "notes": self.notes,
        }


class VocabularyService:
    """Domain service for vocabulary operations.

    Pure Python, no Qt dependencies. All operations return results
    or raise specific exceptions - never silent failures.
    """

    def __init__(
        self,
        vocab: dict[str, Any],
        categories: dict[str, list[str]],
        normalize_jy: Callable[[str], str] | None = None,
    ):
        """
        Args:
            vocab: Vocabulary data structure (mutable, owned by caller)
            categories: Category memberships (mutable, owned by caller)
            normalize_jy: Optional Jyutping normalizer
        """
        self._vocab = vocab
        self._categories = categories
        self._normalize_jy = normalize_jy or self._default_normalize

    @staticmethod
    def _default_normalize(jy: str) -> str:
        """Default Jyutping normalization."""
        return " ".join((jy or "").strip().lower().split())

    def validate_jyutping(self, jyutping: str) -> str:
        """Validate and normalize Jyutping.

        Args:
            jyutping: Raw Jyutping input

        Returns:
            Normalized Jyutping

        Raises:
            JyutpingValidationError: If validation fails
        """
        jy = (jyutping or "").strip()
        if not jy:
            raise JyutpingValidationError("", "Jyutping is empty")

        ok, reason = validate_jyut_syllables(jy)
        if not ok:
            raise JyutpingValidationError(jy, reason or "Invalid format")

        return self._normalize_jy(jy)

    def validate_entry(
        self,
        jyutping: str,
        hanzi: str,
        meanings: list[str] | str,
        categories: list[str] | str,
    ) -> VocabEntry:
        """Validate all entry fields.

        Returns:
            Validated VocabEntry

        Raises:
            ValidationError: If any field is invalid
        """
        # Jyutping
        jy_norm = self.validate_jyutping(jyutping)

        # Hanzi
        hz = (hanzi or "").strip()
        if not hz:
            raise ValidationError("hanzi", hz, "Hanzi is required")

        # Meanings
        if isinstance(meanings, str):
            mn_list = [m.strip() for m in meanings.split(",") if m.strip()]
        else:
            mn_list = [str(m).strip() for m in (meanings or []) if str(m).strip()]

        if not mn_list:
            raise ValidationError("meanings", str(meanings), "At least one meaning required")

        # Categories
        if isinstance(categories, str):
            cat_list = [categories.strip()] if categories.strip() else []
        else:
            cat_list = [str(c).strip() for c in (categories or []) if str(c).strip()]

        # Allow empty categories (will default to 'unassigned')
        # But validate that provided categories are not reserved
        for cat in cat_list:
            if cat.lower() in ("all",):
                raise ValidationError(
                    "category",
                    cat,
                    "Reserved category name"
                )

        return VocabEntry(
            jyutping=jy_norm,
            hanzi=hz,
            meanings=mn_list,
            categories=cat_list or ["unassigned"],
        )

    def check_duplicate_jyutping(self, jyutping: str) -> bool:
        """Check if Jyutping already exists.

        Returns:
            True if duplicate exists
        """
        return is_duplicate_jy(
            jyutping,
            vocab=self._vocab,
            normalize=self._normalize_jy
        )

    def check_exact_duplicate(self, jyutping: str, hanzi: str) -> bool:
        """Check if exact (jyutping, hanzi) pair exists.

        Returns:
            True if exact duplicate exists
        """
        return is_exact_duplicate_entry(
            self._vocab,
            jyutping,
            hanzi,
            normalize=self._normalize_jy
        )

    def add_entry(
        self,
        jyutping: str,
        hanzi: str,
        meanings: list[str] | str,
        categories: list[str] | str,
        allow_duplicate_jy: bool = False,
        notes: str = "",
    ) -> VocabEntry:
        """Add a new vocabulary entry.

        Args:
            jyutping: Jyutping pronunciation
            hanzi: Chinese characters
            meanings: Meanings (list or comma-separated string)
            categories: Categories (list or single string)
            allow_duplicate_jy: If False, raise on duplicate Jyutping
            notes: Optional notes

        Returns:
            Created VocabEntry

        Raises:
            DuplicateEntryError: If entry already exists
            ValidationError: If validation fails
        """
        # Validate all fields
        entry = self.validate_entry(jyutping, hanzi, meanings, categories)

        # Check duplicates
        if not allow_duplicate_jy and self.check_duplicate_jyutping(entry.jyutping):
            raise DuplicateEntryError(entry.jyutping, entry.hanzi)

        if self.check_exact_duplicate(entry.jyutping, entry.hanzi):
            raise DuplicateEntryError(
                entry.jyutping,
                entry.hanzi,
                {"reason": "Exact entry already exists"}
            )

        # Add to vocab (legacy format: hanzi -> [meanings, jyutping])
        self._vocab[entry.hanzi] = (entry.meanings, entry.jyutping)

        # Add to categories
        for cat in entry.categories:
            if cat not in self._categories:
                self._categories[cat] = []
            if entry.hanzi not in self._categories[cat]:
                self._categories[cat].append(entry.hanzi)

        # Remove from 'unassigned' if now in real category
        if len(entry.categories) > 0 and entry.categories != ["unassigned"]:
            if "unassigned" in self._categories:
                try:
                    self._categories["unassigned"].remove(entry.hanzi)
                except ValueError:
                    pass

        return entry

    def update_entry(
        self,
        original_hanzi: str,
        jyutping: str,
        hanzi: str,
        meanings: list[str] | str,
        categories: list[str] | str,
        notes: str = "",
    ) -> VocabEntry:
        """Update an existing vocabulary entry.

        Args:
            original_hanzi: Current Hanzi key (may change)
            jyutping: New Jyutping
            hanzi: New Hanzi
            meanings: New meanings
            categories: New categories
            notes: New notes

        Returns:
            Updated VocabEntry

        Raises:
            ValidationError: If validation fails
        """
        # Validate
        entry = self.validate_entry(jyutping, hanzi, meanings, categories)

        # Remove old entry
        if original_hanzi != hanzi and original_hanzi in self._vocab:
            del self._vocab[original_hanzi]

        # Remove from old categories
        for cat_members in self._categories.values():
            try:
                cat_members.remove(original_hanzi)
            except ValueError:
                pass

        # Add new entry
        self._vocab[entry.hanzi] = (entry.meanings, entry.jyutping)

        # Add to new categories
        for cat in entry.categories:
            if cat not in self._categories:
                self._categories[cat] = []
            if entry.hanzi not in self._categories[cat]:
                self._categories[cat].append(entry.hanzi)

        return entry


__all__ = [
    "VocabEntry",
    "VocabularyService",
]

File: flutter_app/lib/domain/vocabulary_service.dart
import 'exceptions.dart';
import 'jyutping_validation.dart';
import 'duplicate_rules.dart';

class VocabEntry {
  final String jyutping;
  final String hanzi;
  final List<String> meanings;
  final List<String> categories;
  final String notes;

  const VocabEntry({
    required this.jyutping,
    required this.hanzi,
    required this.meanings,
    required this.categories,
    this.notes = '',
  });

  Map<String, dynamic> toMap() {
    return {
      'jyutping': jyutping,
      'hanzi': hanzi,
      'meaning': meanings.join(', '),
      'gloss': meanings.join(', '),
      'categories': List<String>.from(categories),
      'category': categories.isNotEmpty ? categories.first : '',
      'notes': notes,
    };
  }
}

class VocabularyService {
  final Map<String, dynamic> _vocab;
  final Map<String, List<String>> _categories;
  final String Function(String) _normalizeJy;

  VocabularyService({
    required Map<String, dynamic> vocab,
    required Map<String, List<String>> categories,
    String Function(String)? normalizeJy,
  })  : _vocab = vocab,
        _categories = categories,
        _normalizeJy = normalizeJy ?? _defaultNormalize;

  static String _defaultNormalize(String jy) {
    return jy.trim().toLowerCase().split(RegExp(r'\s+')).where((p) => p.isNotEmpty).join(' ');
  }

  String validateJyutping(String jyutping) {
    final jy = jyutping.trim();
    if (jy.isEmpty) {
      throw JyutpingValidationError('', 'Jyutping is empty');
    }
    final res = validateJyutSyllables(jy);
    final ok = res[0] as bool;
    final reason = res[1] as String?;
    if (!ok) {
      throw JyutpingValidationError(jy, reason ?? 'Invalid format');
    }
    return _normalizeJy(jy);
  }

  VocabEntry validateEntry({
    required String jyutping,
    required String hanzi,
    required dynamic meanings,
    required dynamic categories,
  }) {
    final jyNorm = validateJyutping(jyutping);

    final hz = hanzi.trim();
    if (hz.isEmpty) {
      throw ValidationError('hanzi', hz, 'Hanzi is required');
    }

    List<String> mnList;
    if (meanings is String) {
      mnList = meanings.split(',').map((m) => m.trim()).where((m) => m.isNotEmpty).toList();
    } else if (meanings is List) {
      mnList = meanings.map((m) => m.toString().trim()).where((m) => m.isNotEmpty).toList();
    } else {
      mnList = [];
    }
    if (mnList.isEmpty) {
      throw ValidationError('meanings', meanings.toString(), 'At least one meaning required');
    }

    List<String> catList;
    if (categories is String) {
      catList = categories.trim().isNotEmpty ? [categories.trim()] : <String>[];
    } else if (categories is List) {
      catList = categories.map((c) => c.toString().trim()).where((c) => c.isNotEmpty).toList();
    } else {
      catList = [];
    }
    for (final cat in catList) {
      if (cat.toLowerCase() == 'all') {
        throw ValidationError('category', cat, 'Reserved category name');
      }
    }

    return VocabEntry(
      jyutping: jyNorm,
      hanzi: hz,
      meanings: mnList,
      categories: catList.isNotEmpty ? catList : ['unassigned'],
    );
  }

  bool checkDuplicateJyutping(String jyutping) {
    return isDuplicateJy(jyutping, vocab: _vocab, normalize: _normalizeJy);
  }

  bool checkExactDuplicate(String jyutping, String hanzi) {
    return isExactDuplicateEntry(_vocab, jyutping, hanzi, normalize: _normalizeJy);
  }

  List<dynamic>? getEntryRaw(String hanzi) {
    final hz = hanzi.trim();
    if (hz.isEmpty) {
      return null;
    }
    final row = _vocab[hz];
    if (row is List && row.isNotEmpty) {
      final meanings = row[0];
      final jy = row.length >= 2 ? row[1] : '';
      return [meanings, jy.toString()];
    }
    return null;
  }

  dynamic getMeaningsRaw(String hanzi) {
    final entry = getEntryRaw(hanzi);
    return entry != null ? entry[0] : null;
  }

  VocabEntry addEntry({
    required String jyutping,
    required String hanzi,
    required dynamic meanings,
    required dynamic categories,
    bool allowDuplicateJy = false,
    String notes = '',
  }) {
    final entry = validateEntry(
      jyutping: jyutping,
      hanzi: hanzi,
      meanings: meanings,
      categories: categories,
    );

    if (!allowDuplicateJy && checkDuplicateJyutping(entry.jyutping)) {
      throw DuplicateEntryError(entry.jyutping, entry.hanzi);
    }
    if (checkExactDuplicate(entry.jyutping, entry.hanzi)) {
      throw DuplicateEntryError(entry.jyutping, entry.hanzi, {'reason': 'Exact entry already exists'});
    }

    _vocab[entry.hanzi] = [entry.meanings, entry.jyutping];

    for (final cat in entry.categories) {
      _categories.putIfAbsent(cat, () => <String>[]);
      if (!_categories[cat]!.contains(entry.hanzi)) {
        _categories[cat]!.add(entry.hanzi);
      }
    }

    if (entry.categories.isNotEmpty && !(entry.categories.length == 1 && entry.categories.first == 'unassigned')) {
      final unassigned = _categories['unassigned'];
      if (unassigned != null) {
        unassigned.remove(entry.hanzi);
      }
    }

    return VocabEntry(
      jyutping: entry.jyutping,
      hanzi: entry.hanzi,
      meanings: entry.meanings,
      categories: entry.categories,
      notes: notes,
    );
  }

  VocabEntry updateEntry({
    required String originalHanzi,
    required String jyutping,
    required String hanzi,
    required dynamic meanings,
    required dynamic categories,
    String notes = '',
  }) {
    final entry = validateEntry(
      jyutping: jyutping,
      hanzi: hanzi,
      meanings: meanings,
      categories: categories,
    );

    if (originalHanzi != hanzi && _vocab.containsKey(originalHanzi)) {
      _vocab.remove(originalHanzi);
    }

    for (final cat in _categories.values) {
      cat.remove(originalHanzi);
    }

    _vocab[entry.hanzi] = [entry.meanings, entry.jyutping];

    for (final cat in entry.categories) {
      _categories.putIfAbsent(cat, () => <String>[]);
      if (!_categories[cat]!.contains(entry.hanzi)) {
        _categories[cat]!.add(entry.hanzi);
      }
    }

    return VocabEntry(
      jyutping: entry.jyutping,
      hanzi: entry.hanzi,
      meanings: entry.meanings,
      categories: entry.categories,
      notes: notes,
    );
  }
}
